### PR TITLE
openlineage: isolate metadata extraction by executing OL methods in separate, forked process

### PR DESCRIPTION
**Pull request by mobuchowski**
This PR builds on https://github.com/apache/airflow/pull/39890 (already merged).

After this change, OpenLineage will execute metadata extraction in separate, forked process. 
It's a technique modeled to what interaction between `LocalTaskJobRunner` and `StandardTaskRunner` looks like - a process, in this case process of `StandardTaskRunner` watches over OpenLineage listener process during metadata extraction. 

This adds a layer of isolation between task execution and OpenLineage, adding a level of assurance that OpenLineage execution does not interfere with task execution in a way other than taking time.
Additionally, this allows us to add configurable timeout for OL execute methods.

The reason for that is, beyond configurability, that sometimes metadata extraction code can hang - for example, when dealing with Snowflake connection issue https://github.com/snowflakedb/snowflake-connector-python/pull/1898 - and we want to give as much guarantees that OL will not cause task to fail.

### CODE REVIEW

File: airflow/providers/openlineage/conf.py [hunk 1]
```diff
@@ -33,7 +33,15 @@
 import os
 from typing import Any
 
-from airflow.compat.functools import cache
+# Disable caching if we're inside tests - this makes config easierg to mock.
```
Comment (potiuk) [inline airflow/providers/openlineage/conf.py]:
```suggestion
# Disable caching if we're inside tests - this makes config easier to mock.
```

Comment (mobuchowski) [reply airflow/providers/openlineage/conf.py]: Fixed.

File: airflow/providers/openlineage/conf.py [hunk 2]
```diff
@@ -33,7 +33,15 @@
 import os
 from typing import Any
 
-from airflow.compat.functools import cache
+# Disable caching if we're inside tests - this makes config easier to mock.
+if os.getenv("PYTEST_VERSION"):
```
Comment (potiuk) [inline airflow/providers/openlineage/conf.py:37]: NICE.

File: airflow/providers/openlineage/plugins/listener.py [hunk 3]
```diff
@@ -319,10 +322,46 @@ def on_failure():
...
+        if pid:
+            process = psutil.Process(pid)
+            try:
+                self.log.debug("Waiting for process %s", pid)
+                process.wait(conf.execution_timeout())
+            except psutil.TimeoutExpired:
+                self.log.warning(
+                    "OpenLineage process %s expired. This should not affect process execution.", pid
+                )
+                process.kill()
```
Comment (potiuk) [inline airflow/providers/openlineage/plugins/listener.py]: I think we should add process.terminate() call as well and get another (possibly hard-coded) timeout before we send kill. Sending SIGKILL should generally be last-resort kinda thing - because it might leave some resources opened, locks locked etc. so we should really only do it if earlier SIGTERM did not terminate the process in reasonable time.

File: airflow/providers/openlineage/plugins/listener.py [hunk 4]
```diff
@@ -319,10 +322,46 @@ def on_failure():
...
+            except BaseException:
+                # Kill the process directly.
+                pass
```
Comment (potiuk) [inline airflow/providers/openlineage/plugins/listener.py]: No need for pass here.

File: airflow/providers/openlineage/plugins/listener.py [hunk 5]
```diff
@@ -319,10 +322,46 @@ def on_failure():
...
+                try:
+                    process.kill()
```
Comment (potiuk) [inline airflow/providers/openlineage/plugins/listener.py]: Here also terminate is a better approach initially I think.

File: airflow/providers/openlineage/sqlparser.py [hunk 6]
```diff
@@ -116,19 +117,26 @@ def from_table_meta(
     return Dataset(namespace=namespace, name=name if not is_uppercase else name.upper())
 
 
-class SQLParser:
+class SQLParser(LoggingMixin):
```
Comment (potiuk) [inline airflow/providers/openlineage/sqlparser.py]: Those changes seem pretty unrelated - could we extrac them to a separate PR maybe?

File: airflow/providers/openlineage/utils/sql.py [hunk 7]
```diff
@@ -33,6 +34,9 @@
     from airflow.hooks.base import BaseHook
 
 
+log = logging.getLogger(__name__)
```
Comment (potiuk) [inline airflow/providers/openlineage/utils/sql.py]: Same here - it seems that while helpful to debug, likely we could have a separate them out - they add a bit noise to the change.

File: airflow/providers/snowflake/hooks/snowflake.py [hunk 8]
```diff
@@ -473,10 +473,10 @@ def get_openlineage_database_specific_lineage(self, _) -> OperatorLineage | None
         from airflow.providers.openlineage.extractors import OperatorLineage
         from airflow.providers.openlineage.sqlparser import SQLParser
 
-        connection = self.get_connection(getattr(self, self.conn_name_attr))
-        namespace = SQLParser.create_namespace(self.get_openlineage_database_info(connection))
-
         if self.query_ids:
+            self.log.info("Getting connector to get database info :sadge:")
```
Comment (potiuk) [inline airflow/providers/snowflake/hooks/snowflake.py]: That one also seem to be loosely related to fork  case, possibly just splitting the PR a bit should help.

**Review by potiuk — COMMENTED** [review]
few nits about splitting the PR

File: airflow/providers/openlineage/sqlparser.py (hunk 6 above)
Comment (potiuk) [reply airflow/providers/openlineage/sqlparser.py]: BTw. One reason why splitting is good idea is that we migh then be able to revert just **fork** change in case we find it causes some problems - other changes look much safer.

**Comment by potiuk** [comment]
Ah I missed those are two commits/PRs already :)

File: airflow/providers/openlineage/plugins/listener.py (hunk 3 above)
Comment (mobuchowski) [reply airflow/providers/openlineage/plugins/listener.py]: Done - now we `process.terminate()` and give it 3 seconds to clean up. I think that's the reasonable timeout, and it should not really be configurable. Only then we try to `process.kill()`.

File: airflow/providers/openlineage/plugins/listener.py (hunk 4 above)
Comment (mobuchowski) [reply airflow/providers/openlineage/plugins/listener.py]: Fixed :)

File: airflow/providers/openlineage/plugins/listener.py (hunk 5 above)
Comment (mobuchowski) [reply airflow/providers/openlineage/plugins/listener.py]: Done as ⏫

File: airflow/providers/openlineage/sqlparser.py (hunk 6 above)
Comment (mobuchowski) [reply airflow/providers/openlineage/sqlparser.py]: Split those into https://github.com/apache/airflow/pull/40200

File: airflow/providers/openlineage/utils/sql.py (hunk 7 above)
Comment (mobuchowski) [reply airflow/providers/openlineage/utils/sql.py]: Same as ⬆️

File: airflow/providers/snowflake/hooks/snowflake.py (hunk 8 above)
Comment (mobuchowski) [reply airflow/providers/snowflake/hooks/snowflake.py]: Same as ⏫ .

**Review by potiuk — APPROVED** [review]
